<a href="https://colab.research.google.com/github/Ananya-AJ/CMPE255-Data-Mining/blob/main/Assignment4/i_Hyperparameter_search_using_keras_tuner_and_data_augmentation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Objective : To find if hyperparameter search of keras tuner improves accuracy of multi class classification.

The dataset used is the CIFAR-10 dataset, which consists of 50,000 training images and 10,000 testing images of different classes (airplane, automobile, bird, cat, deer, dog, frog, horse, ship, and truck). The goal is to train a model that can accurately classify images into one of these 10 categories. 

In [ ]:
!pip install tensorflow-gpu
!pip install keras-tuner

In [4]:
# imports
import tensorflow as tf
from tensorflow import keras
import keras_tuner
from tensorflow.keras import layers
from kerastuner.tuners import RandomSearch
from tensorflow import keras
from tensorflow.keras.datasets import cifar10


<ipython-input-4-a31f5f22114b>:6: DeprecationWarning: `import kerastuner` is deprecated, please use `import keras_tuner`.
  from kerastuner.tuners import RandomSearch


In [49]:
# Load CIFAR-10 dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

# Select a subset of the data for training and testing
train_subset = 15000
test_subset = 500
x_train = x_train[:train_subset]
y_train = y_train[:train_subset]
x_test = x_test[:test_subset]
y_test = y_test[:test_subset]

# Normalize pixel values to be between 0 and 1
x_train = x_train / 255.0
x_test = x_test / 255.0

# Convert labels to one-hot encoding
num_classes = 10
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)


define a simple CNN model with 3 convolutional layers and 2 fully connected layers



In [50]:
def create_model():
    model = keras.Sequential([
        layers.Conv2D(32, (3,3), activation='relu', input_shape=(32,32,3)),
        layers.MaxPooling2D((2,2)),
        layers.Conv2D(64, (3,3), activation='relu'),
        layers.MaxPooling2D((2,2)),
        layers.Conv2D(128, (3,3), activation='relu'),
        layers.Flatten(),
        layers.Dense(128, activation='relu'),
        layers.Dense(num_classes, activation='softmax')
    ])
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model


train the model for 20 epochs and evaluate its accuracy on the test set

In [51]:
# Create and train the model
model = create_model()
history = model.fit(x_train, y_train, epochs=20, validation_data=(x_test, y_test))

# Evaluate the model on the test set
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/20
469/469 [==============================] - 5s 8ms/step - loss: 1.7645 - accuracy: 0.3436 - val_loss: 1.5086 - val_accuracy: 0.4540
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 1.3980 - accuracy: 0.4895 - val_loss: 1.3290 - val_accuracy: 0.5160
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 1.2403 - accuracy: 0.5566 - val_loss: 1.2979 - val_accuracy: 0.5140
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 1.1118 - accuracy: 0.6078 - val_loss: 1.2018 - val_accuracy: 0.5460
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 0.9989 - accuracy: 0.6481 - val_loss: 1.1446 - val_accuracy: 0.5960
Epoch 6/20
469/469 [==============================] - 3s 7ms/step - loss: 0.9022 - accuracy: 0.6811 - val_loss: 1.0898 - val_accuracy: 0.6000
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 0.8092 - accuracy: 0.7135 - val_loss: 1.1085 - val_accuracy: 0.6040
Epoch 

Hyperparameter tuning with Keras Tuner:
We will use the RandomSearch tuner to randomly sample values for the hyperparameters, and we will search over the learning rate, number of filters in the convolutional layers, and number of neurons in the fully connected layers.

In [57]:
from kerastuner.tuners import RandomSearch

# Define the hyperparameter search space
def create_model(hp):
    model = keras.Sequential()
    model.add(layers.Conv2D(hp.Int('filters1', 32, 96, step=32), (3,3), activation='relu', input_shape=x_train.shape[1:]))
    model.add(layers.MaxPooling2D((2,2)))
    model.add(layers.Conv2D(hp.Int('filters2', 32, 96, step=32), (3,3), activation='relu'))
    model.add(layers.MaxPooling2D((2,2)))
    model.add(layers.Conv2D(hp.Int('filters3', 32, 96, step=32), (3,3), activation='relu'))
    model.add(layers.Flatten())
    model.add(layers.Dense(hp.Int('units1', 64, 256, step=32), activation='relu'))
    model.add(layers.Dense(num_classes, activation='softmax'))
    model.compile(optimizer=keras.optimizers.Adam(hp.Choice('learning_rate', [1e-3, 1e-4, 1e-5])),
                  loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [58]:
# Define the tuner
tuner = RandomSearch(
    create_model,
    objective='val_accuracy',
    max_trials=10,
    executions_per_trial=2,
    directory='keras_tuner',
    project_name='cifar10'
)

In [59]:
# Search for the best hyperparameters
tuner.search(x_train, y_train, epochs=30, validation_data=(x_test, y_test))

In [60]:
# Print the best hyperparameters
best_hyperparams = tuner.get_best_hyperparameters()[0]
print('Best hyperparameters:')
print(f'learning_rate = {best_hyperparams.get("learning_rate")}')
print(f'filters1 = {best_hyperparams.get("filters1")}')
print(f'filters2 = {best_hyperparams.get("filters2")}')
print(f'filters3 = {best_hyperparams.get("filters3")}')
print(f'units1 = {best_hyperparams.get("units1")}')

Best hyperparameters:
learning_rate = 0.001
filters1 = 32
filters2 = 64
filters3 = 128
units1 = 192


retrieve the best hyperparameters and create a new model with those values

In [61]:
# Create a new model with the best hyperparameters
model = tuner.hypermodel.build(best_hyperparams)
model.compile(optimizer=keras.optimizers.Adam(best_hyperparams.get('learning_rate')),
              loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model for 20 epochs with the best hyperparameters
history = model.fit(x_train, y_train, epochs=10, validation_data=(x_test, y_test))

# Evaluate the model on the test set with the best hyperparameters
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy with hyperparameter tuning:', test_acc)

Epoch 1/10
469/469 [==============================] - 5s 7ms/step - loss: 1.7371 - accuracy: 0.3524 - val_loss: 1.4866 - val_accuracy: 0.4440
Epoch 2/10
469/469 [==============================] - 3s 6ms/step - loss: 1.3865 - accuracy: 0.4971 - val_loss: 1.2507 - val_accuracy: 0.5540
Epoch 3/10
469/469 [==============================] - 3s 6ms/step - loss: 1.2149 - accuracy: 0.5682 - val_loss: 1.1859 - val_accuracy: 0.5600
Epoch 4/10
469/469 [==============================] - 3s 7ms/step - loss: 1.0940 - accuracy: 0.6182 - val_loss: 1.2590 - val_accuracy: 0.5680
Epoch 5/10
469/469 [==============================] - 3s 7ms/step - loss: 0.9833 - accuracy: 0.6533 - val_loss: 1.1247 - val_accuracy: 0.6100
Epoch 6/10
469/469 [==============================] - 3s 6ms/step - loss: 0.8673 - accuracy: 0.6961 - val_loss: 1.0656 - val_accuracy: 0.6220
Epoch 7/10
469/469 [==============================] - 2s 5ms/step - loss: 0.7811 - accuracy: 0.7279 - val_loss: 1.0705 - val_accuracy: 0.6680
Epoch 

In [62]:
print('Test accuracy with hyperparameter tuning:', test_acc)

Test accuracy with hyperparameter tuning: 0.6679999828338623


After hyper parameter tuning, test accuracy increases to 66% and loss decreases considerably to 1.13 while test accuracy without hyper parameter training was 61% and loss was 2.28

Testing the test accuracy using data augmentation and then tuning its hyper parameters
The pipeline includes random flips, rotations, and zooms to create new variations of the input images and prevent overfitting

In [63]:
# Define the data augmentation pipeline
data_augmentation = keras.Sequential(
    [
        layers.experimental.preprocessing.RandomFlip("horizontal"),
        layers.experimental.preprocessing.RandomRotation(0.1),
        layers.experimental.preprocessing.RandomZoom(0.1),
    ]
)

In [68]:
def build_model1(hp):
    model = keras.Sequential([
        data_augmentation,
        layers.Conv2D(hp.Choice('filters1', values=[32, 64, 96]), kernel_size=3, activation='relu', input_shape=(32, 32, 3)),
        layers.Conv2D(hp.Choice('filters2', values=[32, 64]), kernel_size=3, activation='relu'),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Dropout(0.25),
        layers.Conv2D(hp.Choice('filters3', values=[32, 64, 96]), kernel_size=3, activation='relu'),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Dropout(0.25),
        layers.Flatten(),
        layers.Dense(hp.Choice('units1', values=[128, 160, 192, 224, 256]), activation='relu'),
        layers.Dropout(0.5),
        layers.Dense(10, activation='softmax')
    ])
    model.compile(
        optimizer=keras.optimizers.Adam(learning_rate=hp.Choice('learning_rate', values=[1e-3, 1e-4, 1e-5])),
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )
    
    return model

In [71]:
# Define the hyperparameter search space
tuner_search = RandomSearch(
    build_model1,
    objective='val_accuracy',
    max_trials=2,
    executions_per_trial=2,
    directory='tuner_dir',
    project_name='cifar10_hyperparameter_tuning'
)


In [72]:
# Perform the hyperparameter search
tuner_search.search(x_train, y_train, epochs=30, validation_data=(x_test, y_test))

Trial 2 Complete [00h 04m 16s]
val_accuracy: 0.5959999859333038

Best val_accuracy So Far: 0.6399999856948853
Total elapsed time: 00h 08m 25s


Best val accuracy achieved so far is 63% with 1.1 loss

In [73]:
# Get the best hyperparameters
best_hp = tuner_search.get_best_hyperparameters(num_trials=1)[0]
print(f'Best hyperparameters:\n{best_hp.values}\n')

Best hyperparameters:
{'filters1': 32, 'filters2': 32, 'filters3': 64, 'units1': 192, 'learning_rate': 0.001}



In [75]:
# Build the model with the best hyperparameters
model = tuner_search.hypermodel.build(best_hp)


In [76]:
# Train the model
model.fit(x_train, y_train, epochs=30, validation_data=(x_test, y_test))

Epoch 1/30
469/469 [==============================] - 7s 9ms/step - loss: 1.9848 - accuracy: 0.2679 - val_loss: 1.6961 - val_accuracy: 0.3780
Epoch 2/30
469/469 [==============================] - 3s 7ms/step - loss: 1.6873 - accuracy: 0.3854 - val_loss: 1.5301 - val_accuracy: 0.4380
Epoch 3/30
469/469 [==============================] - 3s 6ms/step - loss: 1.5710 - accuracy: 0.4270 - val_loss: 1.4178 - val_accuracy: 0.5040
Epoch 4/30
469/469 [==============================] - 3s 6ms/step - loss: 1.5027 - accuracy: 0.4585 - val_loss: 1.3414 - val_accuracy: 0.5460
Epoch 5/30
469/469 [==============================] - 4s 8ms/step - loss: 1.4470 - accuracy: 0.4790 - val_loss: 1.3592 - val_accuracy: 0.5220
Epoch 6/30
469/469 [==============================] - 3s 6ms/step - loss: 1.4092 - accuracy: 0.4929 - val_loss: 1.2592 - val_accuracy: 0.5400
Epoch 7/30
469/469 [==============================] - 3s 7ms/step - loss: 1.3777 - accuracy: 0.5082 - val_loss: 1.1664 - val_accuracy: 0.5960
Epoch 

In [77]:
# Evaluate the model on the test set
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f'Test accuracy with hyperparameter tuning and data augmentation: {test_acc}')

16/16 [==============================] - 0s 3ms/step - loss: 1.0918 - accuracy: 0.6240
Test accuracy with hyperparameter tuning and data augmentation: 0.6240000128746033


Test accuracy did not increase much but loss reduced considerably to 1.09 with data augmentation and hyperparameter tuning